controller/src/main/java/org/jboss/as/controller/security/CredentialReference.java

```diff
@@ -169,12 +169,21 @@ private static AttributeMarshaller credentialReferenceAttributeMarshaller() {
             @Override
             public void marshallAsElement(AttributeDefinition attribute, ModelNode credentialReferenceModelNode, boolean marshallDefault, XMLStreamWriter writer) throws XMLStreamException {
                 writer.writeStartElement(attribute.getXmlName());
-                if (credentialReferenceModelNode.hasDefined(clearTextAttribute.getName())) {
-                    clearTextAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                ModelNode crNode = null;
+                if (credentialReferenceModelNode.hasDefined(attribute.getName())) {
+                    crNode = credentialReferenceModelNode.get(attribute.getName());
                 } else {
-                    credentialStoreAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialAliasAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
-                    credentialTypeAttribute.marshallAsAttribute(credentialReferenceModelNode, writer);
+                    crNode = credentialReferenceModelNode;
+                }
+                if (!crNode.isDefined()) {
+                    throw new XMLStreamException(ControllerLogger.ROOT_LOGGER.invalid(attribute.getName()));
+                }
```

Comment:
If this is possible something has been allowed previously that should have been detected and rejected. We shouldn't be doing model validation during xml marshaling.

If this shouldn't be possible and this is the equivalent of an assert, perhaps use an assert.